limitador/src/storage/infinispan/sets.rs

```diff
@@ -0,0 +1,114 @@
+// Infinispan does not support sets. This module provides some functions on top
+// of the Infinispan client to be able to work with sets.
+
+use crate::storage::infinispan::dist_lock;
+use crate::storage::infinispan::response::response_to_string;
+use crate::storage::StorageErr;
+use infinispan::errors::InfinispanError;
+use infinispan::{request, Infinispan};
+use serde_json::json;
+use std::collections::HashSet;
+use std::iter::FromIterator;
+
+pub async fn get(
+    infinispan: &Infinispan,
+    cache_name: impl AsRef<str>,
+    set_key: impl AsRef<str>,
+) -> Result<HashSet<String>, InfinispanError> {
+    let get_entry_response = infinispan
+        .run(&request::entries::get(cache_name, &set_key))
+        .await?;
+
+    let set = if get_entry_response.status() == 404 {
+        HashSet::new()
+    } else {
+        // TODO: handle other errors
+        serde_json::from_str(&response_to_string(get_entry_response).await).unwrap()
+    };
+
+    Ok(set)
+}
+
+// Creates the set if it does not exist.
+pub async fn add(
+    infinispan: &Infinispan,
+    cache_name: impl Into<String>,
+    set_key: impl Into<String>,
+    element: impl Into<String>,
+) -> Result<(), StorageErr> {
+    let cache_name = cache_name.into();
+    let set_key = set_key.into();
+
+    dist_lock::lock(infinispan, &cache_name, &set_key).await?;
```

Comment:
This is now calling for the lock guard type :) - check out [`MutexGuard`](https://doc.rust-lang.org/std/sync/struct.MutexGuard.html) for an example.

You don't need to make it so it guards _data_, but it can guard the obvious bugs that show up in this fn. In particular, when you are done with the fn you want to release the lock if you created it successfuly, but you have at least 3 exit points that won't release a taken lock. Perhaps the issue is transient network error, perhaps it's a permanent reset, but the guard will force the code to try to release the lock when leaving this context.

Same thing applies to `delete`.

Reply:
I don't see how to use MutexGuard in this context.

Reply:
The fn can be something like:
```rust
pub async fn lock(client: ..., cache_name: &str, set_key: &str) -> Result<DistLockGuard> {
    // [...]
    Ok(DistLockGuard {
        client,
        set_key,
        // any other info needed to unlock
    })
}
```
Then `DistLockGuard` should have:
```rust
impl Drop for DistLockGuard {
    fn drop(&mut self) {
        // perform unlock here
    }
}
```
Since you are using async, if the type is `Pin`-ned at any point, you'd want to check out [how to drop it](https://doc.rust-lang.org/std/pin/index.html#drop-implementation).

You can also add an `unlock` fn in this type, and if you do so you should probably have a flag about the state of the lock and prevent it from being called concurrently from other threads, ie `!Sync` (`Send` looks ok unlike mutexes which usually require that the same thread performing the lock should do the unlock).

Reply:
We can't easily make the Drop solution work for now without making some changes in the infinispan-rs lib first.
I've added a new commit (the last one in this PR) that solves the problem in a different way.